api/server/volume.go

```diff
@@ -362,62 +330,85 @@ func (vd *volAPI) volumeSet(w http.ResponseWriter, r *http.Request) {
+		// Only update spec if spec and locator are not nil.
+		vol, err := volumes.Inspect(ctx, &api.SdkVolumeInspectRequest{
+			VolumeId: volumeID,
+		})
+		if err != nil {
+			if !sdk.IsErrorNotFound(err) {
+				vd.sendError(vd.name, method, w, err.Error(), http.StatusBadRequest)
+				return
 			}
-			if err != nil {
-				break
+
+			vd.logRequest(method, string(volumeID)).Infoln("Ignoring unmount/detach action on deleted volume.")
+		} else {
+			updateReq := &api.SdkVolumeUpdateRequest{VolumeId: volumeID}
+			if req.Locator != nil && len(req.Locator.VolumeLabels) > 0 {
+				updateReq.Labels = req.Locator.VolumeLabels
+			}
+			if req.Spec != nil {
+				if err = vd.updateReplicaSpecNodeIPstoIds(req.Spec.ReplicaSet); err != nil {
+					vd.sendError(vd.name, method, w, err.Error(), http.StatusBadRequest)
+					return
+				}
+
+				updateReq.Spec = getVolumeUpdateSpec(req.Spec, vol.GetVolume())
 			}
-		}
 
-		unmountOptions := &api.SdkVolumeUnmountOptions{}
-		if req.Options["DELETE_AFTER_UNMOUNT"] == "true" {
-			unmountOptions.DeleteMountPath = true
+			if _, err := volumes.Update(ctx, updateReq); err != nil {
+				vd.sendError(vd.name, method, w, err.Error(), http.StatusBadRequest)
+				return
+			}
 		}
-		if req.Options["WAIT_BEFORE_DELETE"] == "true" {
-			unmountOptions.NoDelayBeforeDeletingMountPath = false
-		} else {
-			unmountOptions.NoDelayBeforeDeletingMountPath = true
+	}
+
+	if req.Action != nil {
+		if req.Action.IsAttach() {
+			_, err = mountAttachClient.Attach(ctx, &api.SdkVolumeAttachRequest{
+				VolumeId:      volumeID,
+				Options:       attachOptions,
+				DriverOptions: req.GetOptions(),
+			})
+		} else if req.Action.IsDetach() {
+			_, err = mountAttachClient.Detach(ctx, &api.SdkVolumeDetachRequest{
+				VolumeId:      volumeID,
+				Options:       detachOptions,
+				DriverOptions: req.GetOptions(),
+			})
 		}
 
-		if req.Action.Mount != api.VolumeActionParam_VOLUME_ACTION_PARAM_NONE {
-			if req.Action.Mount == api.VolumeActionParam_VOLUME_ACTION_PARAM_ON {
+		if err == nil {
+			if req.Action.IsMount() {
 				if req.Action.MountPath == "" {
 					err = fmt.Errorf("Invalid mount path")
-					break
+				} else {
+					_, err = mountAttachClient.Mount(ctx, &api.SdkVolumeMountRequest{
+						VolumeId:      volumeID,
+						MountPath:     req.Action.MountPath,
+						DriverOptions: req.GetOptions(),
+					})
 				}
-				_, err = mountAttachClient.Mount(ctx, &api.SdkVolumeMountRequest{
-					VolumeId:      volumeID,
-					MountPath:     req.Action.MountPath,
-					DriverOptions: req.GetOptions(),
-				})
-			} else {
+			} else if req.Action.IsUnMount() {
 				_, err = mountAttachClient.Unmount(ctx, &api.SdkVolumeUnmountRequest{
 					VolumeId:      volumeID,
 					MountPath:     req.Action.MountPath,
 					Options:       unmountOptions,
 					DriverOptions: req.GetOptions(),
 				})
 			}
-			if err != nil {
-				break
-			}
 		}
-		break
 	}
 
 	resVol, err2 := volumes.Inspect(ctx, &api.SdkVolumeInspectRequest{
```

Comment:
Don't we need the ErrNotFound error check here?

Reply:
not needed here since inspect failure is saved in `err2`. The actual response in the REST response will be `err` which is for the set calls.